onnx/external_data_helper.py

```diff
@@ -0,0 +1,185 @@
+        if info.offset is not None:
+
+            # Pad file to required offset if needed
+            data_file.seek(0, 2)
+            file_size = data_file.tell()
+            if info.offset > file_size:
+                data_file.write(b"\0" * (info.offset - file_size))
+
+            data_file.seek(info.offset)
+        data_file.write(raw_data)
+
+
+def _get_all_tensors(onnx_model_proto):  # type: (ModelProto) -> Iterable[TensorProto]
+    """Scan an ONNX model for all tensors and return as an iterator."""
+    return chain(_get_initializer_tensors(onnx_model_proto),
+                 _get_attribute_tensors(onnx_model_proto))
+
+
+def _get_initializer_tensors(onnx_model_proto):  # type: (ModelProto) -> Iterable[TensorProto]
+    """Create an iterator of initializer tensors from ONNX model."""
+    for initializer in onnx_model_proto.graph.initializer:
+        yield initializer
+
+
+def _get_attribute_tensors(onnx_model_proto):  # type: (ModelProto) -> Iterable[TensorProto]
+    """Create an iterator of tensors from node attributes of an ONNX model."""
+    for node in onnx_model_proto.graph.node:
+        for attribute in node.attribute:
+            if attribute.HasField("t"):
+                yield attribute.t
+            for tensor in attribute.tensors:
+                yield tensor
+
+
+def _sanitize_path(path):  # type: (Text) -> Text
+    """Remove path components which would allow traversing up a directory tree from a base path.
+
+    Note: This method is currently very basic and should be expanded.
+    """
+    return path.lstrip('/.')
+
+
+def uses_external_data(tensor):  # type: (TensorProto) -> bool
+    """Return true if the tensor stores data in an external location."""
+    return tensor.HasField("data_location") and tensor.data_location == TensorProto.EXTERNAL
+
+
+def remove_external_data_field(tensor, field_key):  # type: (TensorProto, Text) -> None
+    """
+    Remove a field from a Tensor's external_data key-value store.
+
+    Modifies tensor object in place.
+
+    @params
+    tensor: Tensor object from which value will be removed
+    field_key: The key of the field to be removed
+    """
+    for (i, field) in enumerate(tensor.external_data):
+        if field.key == field_key:
+            del tensor.external_data[i]
+
+
+def add_basepath_to_external_data_tensors(model, filepath):  # type: (ModelProto, Text) -> ModelProto
+    """
+    Add basepath value to the external_data field of all tensors in model.
+
+    Base path information is useful for finding the external data files on disk.
+    Modifies model object in place.
+
+    @params
+    model: Model object to modify.
+    filepath: System path to the directory which should be treated as base path for external data.
+
+    @return
+    The modified model object.
+    """
+    for tensor in _get_all_tensors(model):
+        if len(tensor.external_data):
+            tensor.external_data.add()
+            tensor.external_data[-1].key = 'basepath'
```

Comment:
Let's don't add basepath in the external_data, instead, ask user to provide it as an argument.

Reply:
That's not good. It means the model file itself is not self-contained.

Reply:
this is not a valid key in our spec.